In [3]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error, r2_score
#All the important Libraries
import pandas as pd
import numpy as np
from sklearn.preprocessing import OneHotEncoder
#The Data File
df = pd.read_csv('price-range-of-hdb-flats-offered.csv')
# Extract the "room_type" column
room_type = df['room_type']

# Apply one-hot encoding to "room_type"
encoder = OneHotEncoder()
room_type_encoded = encoder.fit_transform(room_type.values.reshape(-1, 1))

# Add the encoded features back to the DataFrame
room_type_encoded_df = pd.DataFrame(room_type_encoded.toarray(), columns=encoder.get_feature_names_out(['room_type']))
rooms = room_type_encoded_df
df = pd.concat([df, rooms], axis=1)

# Drop the original "room_type" column
df = df.drop('room_type', axis=1)

# Save the modified DataFrame to a new CSV file
df.to_csv('encoded_data.csv', index=False)

# Encode the categorical variable(s), if any
df = pd.get_dummies(df, columns=['town'])

# Split the data into training and testing sets
X = df.drop(['max_selling_price'], axis=1)
y = df['max_selling_price']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Scale the data, if needed
# scaler = StandardScaler()
# X_train = scaler.fit_transform(X_train)
# X_test = scaler.transform(X_test)

# Create a Random Forest model
rf = RandomForestRegressor(n_estimators=100, max_depth=10, min_samples_split=5)

# Train the Random Forest model
rf.fit(X_train, y_train)

# Evaluate the performance of the Random Forest model
y_pred = rf.predict(X_test)
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)
print('MSE:', mse)
print('R^2:', r2)

MSE: 238762067.73386234
R^2: 0.9863070886803009


The MSE (Mean Squared Error) value of 236,853,155.40 suggests that on average, the model's predictions are off by approximately $486,7.34 (sqrt(236853155.40)). Whether or not this is a good score depends on the context of the problem and the domain-specific knowledge.

Similarly, the R^2 (coefficient of determination) value of 0.986 indicates that the model explains 98.6% of the variation in the data, which is considered a very good fit. 

In [4]:
# Collect information about the house features
financial_year = 2010
town = 'Ang Mo Kio'
min_selling_price = 120000
max_selling_price_less_ahg_shg = 0
room_type_2room = 0
room_type_3room = 1
room_type_4room = 0
room_type_5room = 0

# Create a DataFrame with the house features
house_features = pd.DataFrame({'financial_year': [financial_year],
                               'town_' + town: [1],
                               'min_selling_price': [min_selling_price],
                               'min_selling_price_less_ahg_shg': [max_selling_price_less_ahg_shg],
                               'room_type_2-room': [room_type_2room],
                               'room_type_3-room': [room_type_3room],
                               'room_type_4-room': [room_type_4room],
                               'room_type_5-room': [room_type_5room]})

# Ensure that the feature names match the names used during training
house_features = house_features.reindex(columns=X_train.columns, fill_value=0)

# Use the trained Random Forest model to predict the maximum selling price
predicted_max_selling_price = rf.predict(house_features)

# Print the predicted maximum selling price
print('Predicted maximum selling price:', predicted_max_selling_price)

Predicted maximum selling price: [166779.24080087]
